### PR TITLE
Fb integrate model

### PR DESCRIPTION
**Pull request by Cell00phane**
This branch updates the current existing code with the model entities implemented during the last sprint, as well as some updates to the entities themselves:
- Date type attributes in entities have been converted to LocalDateTime, and the adapters have been updated to convert between the Firestore Timestamp and the LocalDateTime since there exists no implicit conversion in the code. 
- Items now have an Id as well as an ItemType implemented as an enum
- Users have a birthdate
- UserProfile entity has been added as a separate entity but still needs work to clearly define them conceptually as well as how to work with them in the code.
IMPORTANT: Updated the minimum sdk version to 26 to accomodate LocalDateTime functionality and to avoid methods that require different API levels, which could lead to ambiguous behavior.
- Need to double check when continuing the work on the database interface for the return types and the methods in general

### CODE REVIEW

File: app/src/main/java/com/github/sdpteam15/polyevents/MainActivity.kt [hunk 1]
```diff
@@ -52,30 +49,22 @@ class MainActivity : AppCompatActivity() {
 
         //Set the basic fragment to the home one or to admin hub if it is logged in
         //TODO Add a condition to see if the user is an admin or not and if so, redirect him to the admin hub
-        if(currentUser == null) {
-            HelperFunctions.changeFragment(this, fragments[R.id.ic_home])
-        } else {
-            HelperFunctions.changeFragment(this, fragments[R.id.id_fragment_admin_hub])
-        }
+        redirectAdmin()
```
Comment (AlessioVerardo) [inline app/src/main/java/com/github/sdpteam15/polyevents/MainActivity.kt:52]: Great refactor

File: app/src/main/java/com/github/sdpteam15/polyevents/database/FirestoreDatabaseProvider.kt [hunk 2]
```diff
@@ -203,37 +191,33 @@ object FirestoreDatabaseProvider : DatabaseInterface {
...
+        uid: String?,
+        userAccess: UserEntity?
     ): Observable<Boolean> = thenDoMultGet(
         firestore!!.collection(USER_COLLECTION)
-            .document(uid)
+            .document(uid!!)
             .get()
     ) {
         //TODO once the data class User is created, set the user with the correct value
         user.postValue(userAccess)
```
Comment (AlessioVerardo) [inline app/src/main/java/com/github/sdpteam15/polyevents/database/FirestoreDatabaseProvider.kt:203]: Now you can create a new UserEntity based on the attributes returned by firebase right ? So that the user informations on the profile fragment will be load asynchronously

File: app/src/main/java/com/github/sdpteam15/polyevents/fragments/LoginFragment.kt [hunk 3]
```diff
@@ -98,6 +97,7 @@ class LoginFragment : Fragment() {
                 if (newValue!!) {
                     if (inDbObservable.value!!) {
                         //If already in database redirect
+                            // TODO: pass retrieved user value to the profile fragment to pass the user info
```
Comment (AlessioVerardo) [inline app/src/main/java/com/github/sdpteam15/polyevents/fragments/LoginFragment.kt:100]: I think it's not necessary to get the user information since the fragment will get the info on its own

File: app/src/main/java/com/github/sdpteam15/polyevents/fragments/ProfileFragment.kt [hunk 4]
```diff
@@ -53,9 +52,13 @@ class ProfileFragment : Fragment() {
         userInfoLiveData.observe(this) { userInfo ->
             viewRoot.findViewById<EditText>(R.id.profileName).setText(userInfo!!.name)
```
Comment (AlessioVerardo) [inline app/src/main/java/com/github/sdpteam15/polyevents/fragments/ProfileFragment.kt:53]: I think this should be userInfo!!.displayName instead of name based on the refactor you've done in the UserEntity class

File: app/src/main/java/com/github/sdpteam15/polyevents/fragments/ProfileFragment.kt [hunk 5]
```diff
@@ -53,9 +52,13 @@ class ProfileFragment : Fragment() {
...
             viewRoot.findViewById<EditText>(R.id.profileEmail).setText(userInfo!!.email)
-            //TODO Line for the future when the user class will have all the attributes
-            //viewRoot.findViewById<EditText>(R.id.profileUsernameET).setText(userInfo!!.username)
-            //viewRoot.findViewById<EditText>(R.id.profileBirthdayET).setText(userInfo!!.birthday)
+            viewRoot.findViewById<EditText>(R.id.profileUsernameET).setText(userInfo!!.username)
+
+            val userBirthDate = userInfo.birthDate
+            val birthDateFormatted =
+                if (userBirthDate == null) ""
+                else userBirthDate.format(DateTimeFormatter.ISO_LOCAL_DATE)
```
Comment (AlessioVerardo) [inline app/src/main/java/com/github/sdpteam15/polyevents/fragments/ProfileFragment.kt:60]: Cool feature !

File: app/src/main/java/com/github/sdpteam15/polyevents/model/Event.kt [hunk 6]
```diff
@@ -56,55 +55,41 @@ data class Event(
...
-        } else {
-            return inventory.put(item, amount)
-        }
-    }
+     * Add an item to the event's inventory
+     * @param item the item to add
+     * @return true if the item was successfully added
+      */
+    fun addItem(item: Item): Boolean =
+        inventory.add(item)
```
Comment (AlessioVerardo) [inline app/src/main/java/com/github/sdpteam15/polyevents/model/Event.kt:64]: Why did you remove this function ?

File: app/src/main/java/com/github/sdpteam15/polyevents/model/Event.kt [hunk 7]
```diff
@@ -56,55 +55,41 @@ data class Event(
...
+     * @return true if the item is there
      */
-    fun takeItem(item: Item, amount: Int = 1): Int? {
-        val currentAmount = inventory.getOrDefault(item, 0)
-        if (amount > currentAmount) {
-            throw InsufficientAmountException("There are only $currentAmount of $item available")
-        } else {
-            return inventory.put(item, currentAmount - amount)
-        }
-    }
```
Comment (AlessioVerardo) [inline app/src/main/java/com/github/sdpteam15/polyevents/model/Event.kt:97]: Why did you remove this function ?

**Review by AlessioVerardo — COMMENTED** [review]
Nice and big refactor ! I know it was long but you did a great job. However, I cannot approve the PR for now since the android tests doesn't pass. Also, if you could make changes according to my remarks it will be great :)

File: app/src/main/java/com/github/sdpteam15/polyevents/database/FirestoreDatabaseProvider.kt (hunk 2 above)
Comment (Cell00phane) [reply app/src/main/java/com/github/sdpteam15/polyevents/database/FirestoreDatabaseProvider.kt:203]: Yes you're right! I didn't plan on doing it on this branch. I'll make a new branch for that when working on the other methods of the database that still need to be implemented :)

File: app/src/main/java/com/github/sdpteam15/polyevents/fragments/LoginFragment.kt (hunk 3 above)
Comment (Cell00phane) [reply app/src/main/java/com/github/sdpteam15/polyevents/fragments/LoginFragment.kt:100]: I see thanks!

File: app/src/main/java/com/github/sdpteam15/polyevents/fragments/ProfileFragment.kt (hunk 4 above)
Comment (Cell00phane) [reply app/src/main/java/com/github/sdpteam15/polyevents/fragments/ProfileFragment.kt:53]: Yes this wasn't clear because the text input in the layout had as description "Full Name", which I thought refers to the name of the user. Are we sure it's the display name here that we want and not the name?

File: app/src/main/java/com/github/sdpteam15/polyevents/model/Event.kt (hunk 7 above)
Comment (Cell00phane) [reply app/src/main/java/com/github/sdpteam15/polyevents/model/Event.kt:97]: This was the first version when Item didn't have an Id, just an itemType. It was easier to keep track of how many of each item type an event had using a map. But now items are "unique",  so we add them to a list or a set. This can still be changed.

File: app/src/main/java/com/github/sdpteam15/polyevents/fragments/ProfileFragment.kt (hunk 4 above)
Comment (AlessioVerardo) [reply app/src/main/java/com/github/sdpteam15/polyevents/fragments/ProfileFragment.kt]: We could store the displayName attribute of the Firesbase Authentication as the name of the user and remove the displayName attribute from the UserEntity so that we only have one attribute referring to the name

File: app/src/main/java/com/github/sdpteam15/polyevents/model/Event.kt (hunk 7 above)
Comment (AlessioVerardo) [reply app/src/main/java/com/github/sdpteam15/polyevents/model/Event.kt:97]: Ok, I got it !

**Comment by AlessioVerardo** [comment]
Are all the tests updated ? because we lost 5.5 % coverage. Is it possible to overcome this problem and write more tests ? since there are parts of the code that won't be tested (google map for instance),

File: app/src/main/java/com/github/sdpteam15/polyevents/adapter/ItemAdapter.kt [hunk 8]
```diff
@@ -33,28 +32,29 @@ class ItemAdapter(
...
-        fun bind(item: String) {
+        fun bind(item: Item) {
             btnRemove.setOnClickListener {
-                Database.currentDatabase.removeItem(item)
+                currentDatabase.removeItem(item)
                 // TODO ADD data.postValue(items)
                 notifyDataSetChanged()
             }
-            itemName.text = item
+            itemName.text = item.itemType.itemType + item.itemId
```
Comment (AlessioVerardo) [inline app/src/main/java/com/github/sdpteam15/polyevents/adapter/ItemAdapter.kt:50]: Is it really necessary to add the itemId in the item name ? I don't think so :)

File: app/src/main/java/com/github/sdpteam15/polyevents/database/Database.kt [hunk 9]
```diff
@@ -1,16 +1,19 @@
...
 object Database {
-    private var mutableCurrentDatabase: DatabaseInterface? = null
+    var mutableCurrentDatabase: DatabaseInterface? = null
     var currentDatabase: DatabaseInterface
         get() {
             mutableCurrentDatabase = mutableCurrentDatabase ?: FirestoreDatabaseProvider
             return mutableCurrentDatabase!!
         }
         set(value) {
+            Log.d("Database", "New DB " + value.toString())
```
Comment (AlessioVerardo) [inline app/src/main/java/com/github/sdpteam15/polyevents/database/Database.kt:16]: We should remove this when we will merge no ? :)

File: app/src/main/java/com/github/sdpteam15/polyevents/database/DatabaseInterface.kt [hunk 10]
```diff
@@ -18,7 +17,8 @@ interface DatabaseInterface {
     /**
      * Current user of this database
      */
-    val currentUser: DatabaseUserInterface?
+    val currentUser: UserEntity?
```
Comment (AlessioVerardo) [inline app/src/main/java/com/github/sdpteam15/polyevents/database/DatabaseInterface.kt:20]: I'm not sure if we should put the currentUser in the database interface and not in the user class ;)

File: app/src/main/java/com/github/sdpteam15/polyevents/database/FirestoreDatabaseProvider.kt [hunk 11]
```diff
@@ -162,36 +153,34 @@ object FirestoreDatabaseProvider : DatabaseInterface {
...
-        userAccess: UserInterface
+        userAccess: UserEntity?
     ): Observable<Boolean> = thenDoSet(
         firestore!!.collection(USER_COLLECTION)
             .document(uid)
             .update(newValues as Map<String, Any>)
     )
 
-
+    // TODO: First connection is when registering? Because this way we are overwriting data
```
Comment (AlessioVerardo) [inline app/src/main/java/com/github/sdpteam15/polyevents/database/FirestoreDatabaseProvider.kt:163]: No first connection is called only when the user is not already in the database, there are checks in the login fragment. You can safely remove this comment :)

File: app/src/androidTest/java/com/github/sdpteam15/polyevents/ItemsAdminActivityTest.kt [hunk 12]
```diff
@@ -68,15 +74,13 @@ class ItemsAdminActivityTest {
...
-
+    fun removeButtonRemovesItemFromList() {/*
         onView(withId(R.id.id_recycler_items_list)).perform(
             RecyclerViewActions.actionOnItemAtPosition<EventItemAdapter.ItemViewHolder>(
                 0, TestHelper.clickChildViewWithId(R.id.id_remove_item)
             )
         )
         onView(withId(R.id.id_recycler_items_list))
-            .check(RecyclerViewItemCountAssertion(2))
+            .check(RecyclerViewItemCountAssertion(2))*/
```
Comment (AlessioVerardo) [inline app/src/androidTest/java/com/github/sdpteam15/polyevents/ItemsAdminActivityTest.kt:84]: Is it really necessary to keep this test since all the lines are commented ?

File: app/src/androidTest/java/com/github/sdpteam15/polyevents/AdminHubFragmentTest.kt [hunk 13]
```diff
@@ -8,53 +11,113 @@ import androidx.test.espresso.intent.matcher.IntentMatchers
...
+                description = "Super cool activity !" +
+                        " With a super long description that essentially describes and explains" +
+                        " the content of the activity we are speaking of.",
+                startTime = LocalDateTime.of(2021, 3, 7, 14, 15),
+                organizer = "The Aqua Poney team",
+                zoneName = "Swimming pool"
+            )
+        )
+        return events
+    }
```
Comment (AlessioVerardo) [inline app/src/androidTest/java/com/github/sdpteam15/polyevents/AdminHubFragmentTest.kt:107]: Why should we initialise the events since, if I'm correct, I didn't test anything related to the events in the admin hub tests

Comment (Cell00phane) [reply app/src/androidTest/java/com/github/sdpteam15/polyevents/AdminHubFragmentTest.kt:107]: @Rygone I refer to Robin for this

File: app/src/androidTest/java/com/github/sdpteam15/polyevents/ItemsAdminActivityTest.kt (hunk 12 above)
Comment (Cell00phane) [reply app/src/androidTest/java/com/github/sdpteam15/polyevents/ItemsAdminActivityTest.kt:84]: @Rygone I refer to Robin for this one as well. I think this was the only test left to fix, as part of the UI tests.

File: app/src/main/java/com/github/sdpteam15/polyevents/database/FirestoreDatabaseProvider.kt (hunk 11 above)
Comment (Cell00phane) [reply app/src/main/java/com/github/sdpteam15/polyevents/database/FirestoreDatabaseProvider.kt:163]: Yes sorry just a sanity check. Will remove it.

File: app/src/main/java/com/github/sdpteam15/polyevents/database/Database.kt (hunk 9 above)
Comment (Cell00phane) [reply app/src/main/java/com/github/sdpteam15/polyevents/database/Database.kt:16]: @Rygone sorry Robin I refer to you for this one as well.

File: app/src/main/java/com/github/sdpteam15/polyevents/fragments/ProfileFragment.kt (hunk 4 above)
Comment (Cell00phane) [reply app/src/main/java/com/github/sdpteam15/polyevents/fragments/ProfileFragment.kt]: Ok this should make things easier. So if we remove displayName, the "name that will be displayed" will be by default the user's name.

File: app/src/main/java/com/github/sdpteam15/polyevents/adapter/ItemAdapter.kt (hunk 8 above)
Comment (Cell00phane) [reply app/src/main/java/com/github/sdpteam15/polyevents/adapter/ItemAdapter.kt:50]: We don't "have to", but I think it's better than having multiple items with the same name, it's just a way to differentiate them. What do you think?

Comment (Cell00phane) [reply app/src/main/java/com/github/sdpteam15/polyevents/adapter/ItemAdapter.kt:50]: The thing is that ItemAdapter will display each item individually if I understood correctly, and not group items with the same itemType and put a number beside to denote how many we have. 

Again, there's still some work to be done on Items to clearly define how are we going to work with them.

File: app/src/main/java/com/github/sdpteam15/polyevents/database/DatabaseInterface.kt (hunk 10 above)
Comment (Cell00phane) [reply app/src/main/java/com/github/sdpteam15/polyevents/database/DatabaseInterface.kt:20]: I would much prefer to keep global variables like currentUser and currentProfile in the same place, preferably the db, or another object. What do you think?

**Review by Rygone — APPROVED** [review]

File: app/src/androidTest/java/com/github/sdpteam15/polyevents/AdminHubFragmentTest.kt (hunk 13 above)
Comment (Rygone) [reply app/src/androidTest/java/com/github/sdpteam15/polyevents/AdminHubFragmentTest.kt:107]: currentDatabase.getUpcomingEvents() was called by the homepage before, but now its useless

File: app/src/androidTest/java/com/github/sdpteam15/polyevents/ItemsAdminActivityTest.kt (hunk 12 above)
Comment (Rygone) [reply app/src/androidTest/java/com/github/sdpteam15/polyevents/ItemsAdminActivityTest.kt:84]: This test is implemented by @Feez2403 in his branch so I decided to not delete it but just comment it

File: app/src/main/java/com/github/sdpteam15/polyevents/database/Database.kt (hunk 9 above)
Comment (Rygone) [reply app/src/main/java/com/github/sdpteam15/polyevents/database/Database.kt]: corrected

**Comment by Cell00phane** [comment]
> Are all the tests updated ? because we lost 5.5 % coverage. Is it possible to overcome this problem and write more tests ? since there are parts of the code that won't be tested (google map for instance),

Yes this is normal because a lot of obsolete classes were deleted so the total number of lines decreased. But all new classes have been tested and accounted for.

File: app/src/androidTest/java/com/github/sdpteam15/polyevents/ItemsAdminActivityTest.kt (hunk 12 above)
Comment (AlessioVerardo) [reply app/src/androidTest/java/com/github/sdpteam15/polyevents/ItemsAdminActivityTest.kt:84]: Ok, I think we can leave it like this for the  moment

File: app/src/main/java/com/github/sdpteam15/polyevents/database/DatabaseInterface.kt (hunk 10 above)
Comment (AlessioVerardo) [reply app/src/main/java/com/github/sdpteam15/polyevents/database/DatabaseInterface.kt:20]: It's a point of view. I still think that currentUser has nothing to do with the database. But you prefer like this, we can keep this version :)

**Review by AlessioVerardo — APPROVED** [review]
Thanks for all the discussion. Everything is great you can merge this PR!
